text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+
+### Fast-syncing
+
+Fast syncing, while beneficial in its own right, is also of interest from a storage perspective, as being able to recover information quickly from the network or external sources influences our decisions about what information must be kept.
+
+The "fast" part of fast-syncing, as compared to the status quo, is realized from two factors, namely
+
+* **side-loading** data from outside the network through external sources and
+* **transfering** state, namely trie-store state, which is currently not transferrable at all.
+
+### Side-loading
+
+We create a common trait for Blocks, Deploys or any other transferable object that states how it can be serialized and defines a hash function for its serialized representation, as well as a key to separate the hash space. As an example, a block whose hash is `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2` would be adressable as `blockv1_5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`. The `blockv1` denotes a complete scheme, that is it also pins the serialization format. While it is not necessary to separate the hash value space to avoid collisions, the prefix allows for multiple representations, should the format of a serialization or an internal data structure change at some point. We refer to data structures that satisfy this condition as *objects* from here on out.
+
+**Note**: The defunct CEP3 proposed hashing the serialized representation of an object instead of the object itself, which makes it a lot easier for external stores to handle these, as integrity can be checked without deserializing first. However, hashing the internal object minimizes the impact on the existing codebase instead.
+
+We can then define a generic mechanism for requesting these objects from other nodes. The true value however is in being able to serialize these to static external storage.
+
+### S3, HTTP and read/write, read-only stores
+
+As objects are unchanging, they can be readily be side-loaded from an HTTP-server that acts as a read-only store. As an example, should a node need to obtain the block `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`, it can follow the following steps:
+
+1. Check one or more configured external HTTP stores by calling `GET store/objects/5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`
+1. If the request is a hit, try to download the object, deserialize it and validate the hash. Should any of these operations fail, discard and continue. Otherwise return.
+1. Ask the network (typically a node is known that is required to have it) for the object in the same manner.
+
+In the same manner, an S3-compatible API can be employed. Since many S3 providers allow making stored objects public, it would be highly recommendable to use a similar structure for the keys, as a result, every object uploaded to S3 automatically becomes read-only fetchable by other nodes that have access to the S3 bucket in real-time.
```

Comment:
> it would be highly recommendable to use a similar structure for the keys
There's no need for the p2p messages to have URL-friendly keys.  It's sufficient to be able to map between these easily.

Reply:
"Structure" is meant to at least mean "tag + hash" here, not necessarily the hex-encoding. It could be encoded binary as well, I am not entirely sure of the savings though, especially if stream compression is ever to be added.